crmd/notify.c

```diff
@@ -20,154 +20,721 @@
+                if (format[fmt_pos+fmt_len] != 0) {
+                    continue;
+                }
+                fmt_pos = scanned_pos; /* print till end */
+            }
+        } else {
+            scanned_pos = strlen(format);
+            fmt_pos = scanned_pos; /* print till end */
+        }
+        tmp_fmt_s = strndup(&format[printed_pos], fmt_pos - printed_pos);
+#ifdef GCC_FORMAT_NONLITERAL_CHECKING_ENABLED
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wformat-nonliteral"
+#endif
+        date_len += strftime(&date_s[date_len], max-date_len, tmp_fmt_s, &tm);
+#ifdef GCC_FORMAT_NONLITERAL_CHECKING_ENABLED
+#pragma GCC diagnostic pop
+#endif
+        printed_pos = scanned_pos;
+        free(tmp_fmt_s);
+        if (nano_digits) {
+#ifdef GCC_FORMAT_NONLITERAL_CHECKING_ENABLED
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wformat-nonliteral"
+#endif
+            date_len += snprintf(&date_s[date_len], max-date_len, 
+                                 nanofmt_s, nano_s);
+#ifdef GCC_FORMAT_NONLITERAL_CHECKING_ENABLED
+#pragma GCC diagnostic pop
+#endif 
+            nano_digits = 0;
+        }
+    }
+
+    return (date_len == 0)?NULL:strdup(date_s);
 }
 
+/*
+ * end of possibly generic time-handling stuff
+ */
+
+
+/*
+ * syncronize local data with cib
+ */
+
 static void
-set_notify_key(const char *name, const char *cvalue, char *value)
+free_environment_variable_entry(environment_variable_t *entry)
 {
-    int lpc;
-    bool found = 0;
+    free(entry->name);
+    free(entry->value);
+    free(entry);
+}
+
+static void 
+free_notify_list_entry(notify_entry_t *entry)
+{
+    free(entry->id);
+    free(entry->path);
+    free(entry->tstamp_format);
+    free(entry->recipient);
+    if (entry->environment_variables) {
+        g_list_free_full(entry->environment_variables, 
+                         (GDestroyNotify) free_environment_variable_entry);
+    }
+    free(entry);
+}
+
+static void 
+free_notify_list()
+{
+    if (notify_list) {
+        g_list_free_full(notify_list, (GDestroyNotify) free_notify_list_entry);
+        notify_list = NULL;
+    }
+}
+
+static gpointer
+copy_environment_variable_entry(environment_variable_t * src,
+                                gpointer data)
+{
+    environment_variable_t *dst = calloc(1, sizeof(environment_variable_t));
+
+    CRM_ASSERT(dst);
+    dst->name = strdup(src->name);
+    dst->value = src->value?strdup(src->value):NULL;
+    return (gpointer) dst;
+}
+
+static GListPtr
+add_dup_envvar(GListPtr environment_variable_list, 
+               environment_variable_t *entry)
+{
+    environment_variable_t *new_entry = 
+        (environment_variable_t *) calloc(1, sizeof(environment_variable_t));
+
+    CRM_ASSERT(new_entry);
+    *new_entry = (environment_variable_t) {
+        .name = strdup(entry->name),
+        .value = entry->value?strdup(entry->value):NULL
+    };
+    return g_list_prepend(environment_variable_list, new_entry);
```

Comment:
add_dup_envvar was there first ;-)
a lot of room for refactoring in a lot of places - aim of this commit was to nail down the interface